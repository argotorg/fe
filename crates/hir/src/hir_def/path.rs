use super::{kw, IdentId};
use crate::{hir_def::Partial, HirDb};

#[salsa::interned]
pub struct PathId {
    #[return_ref]
    pub segments: Vec<Partial<IdentId>>,
}

impl PathId {
    pub fn last_segment(self, db: &dyn HirDb) -> Partial<IdentId> {
        self.segments(db).last().copied().unwrap_or_default()
    }

    pub fn len(self, db: &dyn HirDb) -> usize {
        self.segments(db).len()
    }

    pub fn is_ident(self, db: &dyn HirDb) -> bool {
        self.len(db) == 1
    }

    pub fn self_ty(db: &dyn HirDb) -> Self {
        let self_ty = Partial::Present(kw::SELF_TY);
        Self::new(db, vec![self_ty])
    }

    pub fn from_ident(db: &dyn HirDb, ident: IdentId) -> Self {
        Self::new(db, vec![Partial::Present(ident)])
    }
}
