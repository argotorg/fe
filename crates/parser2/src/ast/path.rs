use rowan::ast::{support, AstNode};

use super::{ast_node, AstChildren};
use crate::{syntax_node::SyntaxToken, SyntaxKind as SK};

ast_node! {
    /// A path.
    /// `foo::bar::baz`
    pub struct Path,
    SK::Path,
    IntoIterator<Item=PathSegment>,
}
impl Path {
    /// Returns the segments of the path.
    pub fn segments(&self) -> AstChildren<PathSegment> {
        support::children(self.syntax())
    }
}

ast_node! {
    /// A path segment.
    pub struct PathSegment,
    SK::PathSegment
}
impl PathSegment {
    /// Returns the identifier of the segment.
    pub fn ident(&self) -> Option<SyntaxToken> {
        support::token(self.syntax(), SK::Ident)
    }

    /// Returns `true` if the segment  is a `self` keyword.
    pub fn is_self(&self) -> bool {
        support::token(self.syntax(), SK::SelfKw).is_some()
    }

    /// Returns `true` if the segment is a `Self` keyword.
    pub fn is_self_ty(&self) -> bool {
        support::token(self.syntax(), SK::SelfTypeKw).is_some()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        lexer::Lexer,
        parser::{path::PathScope, Parser},
    };

    use wasm_bindgen_test::wasm_bindgen_test;

    fn parse_path(source: &str) -> Path {
        let lexer = Lexer::new(source);
        let mut parser = Parser::new(lexer);
        parser.parse(PathScope::default(), None);
        Path::cast(parser.finish().0).unwrap()
    }

    #[test]
    #[wasm_bindgen_test]
    fn path_ast() {
        let source = r#"self::Foo"#;
        let path = parse_path(source);
        let mut segments = path.segments();

        assert!(segments.next().unwrap().is_self());
        assert_eq!(segments.next().unwrap().ident().unwrap().text(), "Foo");
        assert!(segments.next().is_none());
    }

    #[test]
    #[wasm_bindgen_test]
    fn path_ast2() {
        let source = r#"Self::Dep"#;
        let path = parse_path(source);
        let mut segments = path.segments();

        assert!(segments.next().unwrap().is_self_ty());
        assert_eq!(segments.next().unwrap().ident().unwrap().text(), "Dep");
        assert!(segments.next().is_none());
    }
}
