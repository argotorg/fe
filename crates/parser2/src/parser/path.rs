use crate::SyntaxKind;

use super::{define_scope, token_stream::TokenStream, Parser};

define_scope! {
    #[doc(hidden)]
    pub PathScope,
    Path,
    Inheritance
}
impl super::Parse for PathScope {
    fn parse<S: TokenStream>(&mut self, parser: &mut Parser<S>) {
        parser.set_newline_as_trivia(false);
        parser.parse(PathSegmentScope::default(), None);
        while parser.current_kind() == Some(SyntaxKind::Colon2) {
            parser.bump_expected(SyntaxKind::Colon2);
            parser.parse(PathSegmentScope::default(), None);
        }
    }
}

define_scope! {
    PathSegmentScope,
    PathSegment,
    Inheritance
}
impl super::Parse for PathSegmentScope {
    fn parse<S: TokenStream>(&mut self, parser: &mut Parser<S>) {
        match parser.current_kind() {
            Some(kind) if is_path_segment(kind) => {
                parser.bump();
            }
            _ => parser.error_and_recover("expected path segment", None),
        }
    }
}

pub(super) fn is_path_segment(kind: SyntaxKind) -> bool {
    matches!(
        kind,
        SyntaxKind::SelfType | SyntaxKind::SelfKw | SyntaxKind::Ident
    )
}
