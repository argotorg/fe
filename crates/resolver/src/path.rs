use camino::Utf8PathBuf;
use serde::Deserialize;

use crate::remote::GitDescription;

#[derive(Deserialize, Hash, Debug, Clone, PartialEq, Eq)]
pub struct PathDescription {
    pub path: String,
}

#[derive(Clone, Debug)]
enum PathHead {
    Local(Utf8PathBuf),
    Remote(GitDescription),
}

#[derive(Clone, Debug)]
pub struct FullPathDescription {
    head: PathHead,
    sub_paths: Vec<PathDescription>,
}

impl FullPathDescription {
    pub fn new_local(local_path: &Utf8PathBuf) -> Self {
        FullPathDescription {
            head: PathHead::Local(local_path.clone()),
            sub_paths: vec![],
        }
    }

    pub fn new_remote(git_description: &GitDescription) -> Self {
        FullPathDescription {
            head: PathHead::Remote(git_description.clone()),
            sub_paths: vec![],
        }
    }

    pub fn path(&self) -> Utf8PathBuf {
        match &self.head {
            PathHead::Local(local_path) => {
                self.sub_paths
                    .iter()
                    .fold(local_path.clone(), |acc, path_description| {
                        acc.join(Utf8PathBuf::from(path_description.path.to_owned()))
                            .canonicalize_utf8()
                            .unwrap()
                    })
            }
            PathHead::Remote(git_description) => self.sub_paths.iter().fold(
                git_description.relative_path(),
                |acc, path_description| {
                    acc.join(Utf8PathBuf::from(path_description.path.to_owned()))
                        .canonicalize_utf8()
                        .unwrap()
                },
            ),
        }
    }

    pub fn push_sub_path(&self, path_description: &PathDescription) -> Self {
        let mut sub_paths = self.sub_paths.clone();
        sub_paths.push(path_description.clone());
        Self {
            head: self.head.clone(),
            sub_paths,
        }
    }
}
