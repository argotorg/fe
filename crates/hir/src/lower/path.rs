use fe_parser2::ast;

use crate::{
    hir_def::{IdentId, MaybeInvalid, PathId, PathSegment},
    HirDb,
};

impl PathId {
    pub(crate) fn from_ast(db: &dyn HirDb, ast: ast::Path) -> Self {
        let mut segments = Vec::new();
        for seg in ast.into_iter() {
            let segment = if seg.is_self() {
                MaybeInvalid::Valid(PathSegment::Self_)
            } else if seg.is_self_ty() {
                MaybeInvalid::Valid(PathSegment::SelfTy)
            } else if let Some(ident) = seg.ident() {
                MaybeInvalid::Valid(PathSegment::Ident(IdentId::new(
                    db,
                    ident.text().to_string(),
                )))
            } else {
                MaybeInvalid::invalid()
            };
            segments.push(segment);
        }

        Self::new(db, segments)
    }

    pub(crate) fn maybe_from_ast(db: &dyn HirDb, ast: Option<ast::Path>) -> MaybeInvalid<Self> {
        ast.map(|ast| Self::from_ast(db, ast)).into()
    }
}
