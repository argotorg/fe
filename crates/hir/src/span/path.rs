use parser::ast;

use super::define_lazy_span_item;

define_lazy_span_item!(
    LazyPathSpan,
    ast::Path,
    @idx {
        (segment, LazyPathSegmentSpan),
    }
);

define_lazy_span_item!(LazyPathSegmentSpan);
