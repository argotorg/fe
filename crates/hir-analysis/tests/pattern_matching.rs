mod test_db;
use std::path::Path;

use dir_test::{dir_test, Fixture};
use test_db::initialize_analysis_pass;
use test_db::HirAnalysisTestDb;

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/pattern_matching",
    glob: "exhaustive/*.fe"
)]
fn exhaustive_matches(fixture: Fixture<&str>) {
    let mut db = HirAnalysisTestDb::default();
    let path = Path::new(fixture.path());
    let file_name = path.file_name().and_then(|file| file.to_str()).unwrap();
    let file = db.new_stand_alone(file_name.into(), fixture.content());
    let (top_mod, _) = db.top_mod(file);
    db.assert_no_diags(top_mod);
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/pattern_matching",
    glob: "non_exhaustive/*.fe"
)]
fn non_exhaustive_matches(fixture: Fixture<&str>) {
    let mut db = HirAnalysisTestDb::default();
    let path = Path::new(fixture.path());
    let file_name = path.file_name().and_then(|file| file.to_str()).unwrap();
    let file = db.new_stand_alone(file_name.into(), fixture.content());
    let (top_mod, _) = db.top_mod(file);

    let mut manager = initialize_analysis_pass();

    let diags = manager.run_on_module(&db, top_mod);

    // Verify we have at least one diagnosis
    assert!(
        !diags.is_empty(),
        "Expected pattern matching errors but found none"
    );
}

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/pattern_matching",
    glob: "unreachable/*.fe"
)]
fn unreachable_patterns(fixture: Fixture<&str>) {
    let mut db = HirAnalysisTestDb::default();
    let path = Path::new(fixture.path());
    let file_name = path.file_name().and_then(|file| file.to_str()).unwrap();
    let file = db.new_stand_alone(file_name.into(), fixture.content());
    let (top_mod, _) = db.top_mod(file);

    let mut manager = initialize_analysis_pass();

    let diags = manager.run_on_module(&db, top_mod);

    // Verify we have at least one diagnosis
    assert!(
        !diags.is_empty(),
        "Expected unreachable pattern errors but found none"
    );
}
