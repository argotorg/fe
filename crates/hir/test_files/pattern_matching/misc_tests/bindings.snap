---
source: crates/hir-analysis/tests/pattern_matching.rs
expression: diagnostic_output
input_file: test_files/pattern_matching/misc_tests/bindings.fe
---
Misc test file bindings.fe has 12 diagnostic(s):

error[8-0014]: type must be known here
   ┌─ bindings.fe:35:47
   │
35 │         (Option::Some(a), Option::Some(b)) => a + b
   │                                               ^ type must be known here

error[8-0014]: type must be known here
   ┌─ bindings.fe:53:47
   │
53 │         (Option::Some(a), Option::Some(b)) => a + b
   │                                               ^ type must be known here

error[8-0031]: type annotation is needed
   ┌─ bindings.fe:33:31
   │
33 │     let pair = (Option::None, Option::None)
   │                               ^^^^^^^^^^^^
   │                               │
   │                               type annotation is needed
   │                               consider giving `: Option<_>` here

error[8-0031]: type annotation is needed
   ┌─ bindings.fe:37:10
   │
37 │         (Option::None, Option::Some(b)) => b      // Does 'b' bind correctly?
   │          ^^^^^^^^^^^^
   │          │
   │          type annotation is needed
   │          consider giving `: Option<_>` here

error[8-0031]: type annotation is needed
   ┌─ bindings.fe:38:10
   │
38 │         (Option::None, Option::None) => 0
   │          ^^^^^^^^^^^^
   │          │
   │          type annotation is needed
   │          consider giving `: Option<_>` here

error[8-0031]: type annotation is needed
   ┌─ bindings.fe:38:24
   │
38 │         (Option::None, Option::None) => 0
   │                        ^^^^^^^^^^^^
   │                        │
   │                        type annotation is needed
   │                        consider giving `: Option<_>` here

error[8-0031]: type annotation is needed
   ┌─ bindings.fe:51:31
   │
51 │     let pair = (Option::None, Option::Some(42))
   │                               ^^^^^^^^^^^^^^^^
   │                               │
   │                               type annotation is needed
   │                               consider giving `: Option<{integer}>` here

error[8-0031]: type annotation is needed
   ┌─ bindings.fe:55:10
   │
55 │         (Option::None, Option::Some(b)) => b      // Does 'b' bind when first type unknown?
   │          ^^^^^^^^^^^^
   │          │
   │          type annotation is needed
   │          consider giving `: Option<_>` here

error[8-0031]: type annotation is needed
   ┌─ bindings.fe:56:10
   │
56 │         (Option::None, Option::None) => 0
   │          ^^^^^^^^^^^^
   │          │
   │          type annotation is needed
   │          consider giving `: Option<_>` here

error[8-0031]: type annotation is needed
   ┌─ bindings.fe:56:24
   │
56 │         (Option::None, Option::None) => 0
   │                        ^^^^^^^^^^^^
   │                        │
   │                        type annotation is needed
   │                        consider giving `: Option<_>` here

error[8-0031]: type annotation is needed
    ┌─ bindings.fe:185:9
    │
185 │     let result = Result::Ok(100)
    │         ^^^^^^
    │         │
    │         type annotation is needed
    │         consider giving `: Result<i32, _>` here

error[8-0031]: type annotation is needed
    ┌─ bindings.fe:211:9
    │
211 │     let outer = Result::Ok(Option::None)
    │         ^^^^^
    │         │
    │         type annotation is needed
    │         consider giving `: Result<Option<i32>, _>` here
